### PR TITLE
CR-1126301 ASTeR TC18.01 - xbutil validate dma test - *ERROR* object creation failed user_flags 31, size 0x1000000

### PR DESCRIPTION
**Pull request by AShivangi**
#### Problem solved by the commit
Don't allocate 4GB as the bank can be smaller than that

#### How problem was solved, alternative solutions (if any) and why they were rejected
Instead of blindly allocating 4GB, we allocate the total size of the bank

#### Risks (if any) associated the changes in the commit
None

#### What has been tested and how, request additional testing if necessary
Tested locally and checked the dmesg

#### Documentation impact (if any)
None

### CODE REVIEW

File: src/runtime_src/core/pcie/common/dmatest.h [hunk 1]
```diff
@@ -117,7 +117,7 @@ namespace xcldev {
         DMARunner(xclDeviceHandle handle, size_t size, unsigned flags = 0, size_t totalSize = 0) :
```
Comment (rozumx) [inline src/runtime_src/core/pcie/common/dmatest.h:117]: Please update the copyright date.

File: src/runtime_src/core/pcie/common/dmatest.h [hunk 2]
```diff
@@ -117,7 +117,7 @@ namespace xcldev {
...
                 mHandle(handle),
                 mSize(size),
-                mTotalSize(totalSize ? totalSize : 0x100000000),
```
Comment (rozumx) [inline src/runtime_src/core/pcie/common/dmatest.h:120]: Thank you for removing this code.  While simple, it is wrong on many levels.

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 3]
```diff
@@ -1002,10 +991,16 @@ dmaTest(const std::shared_ptr<xrt_core::device>& _dev, boost::property_tree::ptr
     logger(_ptTest, "Details", (boost::format("Buffer size - '%s'") % xrt_core::utils::unit_convert(block_size)).str());
 
     // check if the bank has enough memory to allocate
-    //  m_size is in KB so convert block_size (bytes) to KB for comparision
+    // m_size is in KB so convert block_size (bytes) to KB for comparison
     if(mem.m_size < (block_size/1024))
       continue;
 
+    size_t totalSize = 0;
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:998]: This code is functional and works for this release.  
Going forward, you might wish to refactore the code and add a set of helper routines in a common header file that will perform a cleaner conversion.

I came across the following code when reviewing a PR from Soren:
`constexpr size_t operator"" _kb(unsigned long long v)  { return 1024u * v; }`
and it is used as follows:
`constexpr auto default_address_range = 64_kb;`

I can imagine a set of these helper routines; _KB, _MB, _GB, _TB

Again something to think about.

**Review by rozumx — APPROVED** [review]

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp (hunk 3 above)
Comment (AShivangi) [reply src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:998]: `64_kb` means 64 kb converted into bytes?

Comment (AShivangi) [reply src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:998]: Also, we have `XBU::string_to_bytes("2G")` which is pretty readable

**Comment by AShivangi** [comment]
retest this please

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp (hunk 3 above)
Comment (stsoe) [reply src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:998]: But very arcane.   `2_gb` is more readable and much more efficient.

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 4]
```diff
@@ -1002,10 +991,16 @@ dmaTest(const std::shared_ptr<xrt_core::device>& _dev, boost::property_tree::ptr
...
+    if (xrt_core::device_query<xrt_core::query::pcie_vendor>(_dev) == ARISTA_ID)
+      totalSize = 0x20000000; // 512 MB 
+    else
+      totalSize = std::min((mem.m_size * 1024), XBU::string_to_bytes("2G")); // minimum of mem size in bytes and 2 GB
```
[the remark this answers is not in the document]
Comment (AShivangi) [reply src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:1002]: Done, I'll fix it in the next release: [CR-1127118](https://jira.xilinx.com/browse/CR-1127118)

**Pull request merged**